flask_monitoringdashboard/database/data_base_queries/mongo_db_objects.py

```diff
@@ -26,19 +27,59 @@ def _safe_mongo_call(*args, **kwargs):
     return _safe_mongo_call
 
 
+def safe_mongo_call_find(call, call_type):
+    """
+    In order to minimize the risk of failure and the impact on the app performance,
+    we need to allow the application to be able to read from the secondaries.
+    In such cases, we will have an eventual data consistence. Hence, we should re-try to read
+    """
+    def has_content(result):
+        if call_type == "find_one":
+            # we will have content is the object is not None
+            return result is not None
+        if call_type == "find":
+            return result and len(list(result.clone())) > 0
+
+    def _safe_mongo_call_find(*args, **kwargs):
+        for i in range(5):
+            try:
+                for _i in range(6):
+                    result = call(*args, **kwargs)
+                    if _i == 5 or has_content(result):
+                        return result
+                    else:
+                        time.sleep(pow(2, i))
+            except (AutoReconnect, ServerSelectionTimeoutError):
+                time.sleep(pow(2, i))
+        else:
+            raise AutoReconnect()
+
+    return _safe_mongo_call_find
+
+
+def do_nothing_function(*args, **kwargs): pass
+
+
 class CollectionWrapper:
     def __init__(self, collection):
         self.collection = collection
 
     def __getattr__(self, item):
         elem = getattr(self.collection, item)
         if callable(elem):
-            return safe_mongo_call(elem)
+            if item in ["create_index", "drop_indexes"] and \
+                    os.environ.get("MONITORING_DISABLED_INDEX_CREATION") == "true":
+                return do_nothing_function
+            if item in ["find", "find_one"]:
+                return safe_mongo_call_find(elem, item)
+            else:
+                return safe_mongo_call(elem)
 
 
 class Base(dict):
     def __init__(self, new_content=None):
         new_content.pop("_id", None)
+        new_content["__creation_datetime__"] = datetime.datetime.utcnow()
```

Comment:
what is this for?

Reply:
Data retention. We should be able to decide if we want to clean or reduce the amount of data and having a creation data can help